ir/instr.cpp

```diff
@@ -1865,4 +1865,121 @@ unique_ptr<Instr> Memcpy::dup(const string &suffix) const {
 }
 
+vector<Value*> ExtractElement::operands() const {
+  return { a, idx };
+}
+
+void ExtractElement::rauw(const Value &what, Value &with) {
+  RAUW(a);
+  RAUW(idx);
+}
+
+void ExtractElement::print(ostream &os) const {
+  os << getName() << " = extractelement " << *a << ", " << *idx;
+}
+
+StateValue ExtractElement::toSMT(State &s) const {
+  auto &[iv, ip] = s[*idx];
+  auto aty = a->getType().getAsAggregateType();
+  expr inbounds = iv.ult(aty->numElementsConst());
+  auto [rv, rp] = aty->extract(s[*a], iv);
+  return { move(rv), ip && inbounds && rp };
+}
+
+expr ExtractElement::getTypeConstraints(const Function &f) const {
+  return Value::getTypeConstraints() &&
+         a->getType().enforceVectorType([&](auto &ty) { return ty == getType(); }) &&
+         idx->getType().enforceIntType();
+}
+
+unique_ptr<Instr> ExtractElement::dup(const string &suffix) const {
+  return make_unique<ExtractElement>(getType(), getName() + suffix, *a, *idx);
+}
+
+
+vector<Value*> InsertElement::operands() const {
+  return { a, e, idx };
+}
+
+void InsertElement::rauw(const Value &what, Value &with) {
+  RAUW(a);
+  RAUW(e);
+  RAUW(idx);
+}
+
+void InsertElement::print(ostream &os) const {
+  os << getName() << " = insertelement" << *a << ", " << *e << ", " << *idx;
+}
+
+StateValue InsertElement::toSMT(State &s) const {
+  auto &[iv, ip] = s[*idx];
+  auto aty = a->getType().getAsAggregateType();
+  expr inbounds = iv.ult(aty->numElementsConst());
+  auto [rv, rp] = aty->update(s[*a], s[*e], iv);
+  return { move(rv), expr::mkIf(ip && inbounds,
+                                move(rp), expr::mkInt(-1, aty->bits())) };
+}
+
+expr InsertElement::getTypeConstraints(const Function &f) const {
+  return Value::getTypeConstraints() &&
+         getType() == a->getType() &&
+         a->getType().enforceVectorType([&](auto &ty) { return ty == e->getType(); }) &&
+         idx->getType().enforceIntType();
+}
+
+unique_ptr<Instr> InsertElement::dup(const string &suffix) const {
+  return make_unique<InsertElement>(getType(), getName() + suffix, *a, *e, *idx);
+}
+
+
+vector<Value*> ShuffleVector::operands() const {
+  return { a, b, m };
+}
+
+void ShuffleVector::rauw(const Value &what, Value &with) {
+  RAUW(a);
+  RAUW(b);
+  RAUW(m);
+}
+
+void ShuffleVector::print(ostream &os) const {
+  os << getName() << " = shufflevector " << *a << ", " << *b << ", " << *m;
+}
+
+StateValue ShuffleVector::toSMT(State &s) const {
+  auto aty = a->getType().getAsAggregateType();
+  auto mty = m->getType().getAsAggregateType();
+  auto ty = getType().getAsAggregateType();
+  vector<StateValue> vals;
+
+  for (unsigned i = 0, e = mty->numElementsConst(); i!= e; ++i) {
+    auto [iv, ip] = mty->extract(s[*m], i);
+    expr sz = expr::mkUInt(aty->numElementsConst(), 32);
+    auto [av, ap] = aty->extract(s[*a], iv.urem(sz));
+    auto [bv, bp] = aty->extract(s[*b], iv.urem(sz));
+    expr inbounds = iv.ult(sz + sz);
+    expr v = expr::mkIf(iv.ult(sz), move(av), move(bv));
```

Comment:
Please use ule/uge whenever possible (cheaper in Z3)

Reply:
fixed:)